src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,232 @@
+	PG_RETURN_VOID();
+}
+
+
+Datum
+replication_origin_session_end_no_publish(PG_FUNCTION_ARGS)
+{
+	repl_origin_local_session_end_no_publish();
+	PG_RETURN_VOID();
+}
+
+
+Datum
+replication_origin_session_is_no_publish(PG_FUNCTION_ARGS)
+{
+	bool result = repl_origin_local_session_is_no_publish();
+	PG_RETURN_BOOL(result);
+}
+
+
+void
+repl_origin_local_session_start_no_publish(void)
+{
+	originalOriginId = replorigin_session_origin;
+	replorigin_session_origin = DoNotReplicateId;
+}
+
+
+void
+repl_origin_local_session_end_no_publish(void)
+{
+	replorigin_session_origin = originalOriginId;
+}
+
+
+bool
+repl_origin_local_session_is_no_publish(void)
+{
+	return (replorigin_session_origin != InvalidRepOriginId);
+}
+
+
+/*
+ * SetupReplicationOriginRemoteSession sets up a new replication origin session in a
+ * remote session. The identifier is used to create a unique replication origin name
+ * for the session in the remote node.
+ */
+void
+SetupReplicationOriginRemoteSession(MultiConnection *connection)
+{
+	/*elog(LOG, "Setting up remote replication origin session %p %s,%d", connection, connection->hostname, connection->port); */
+	if (connection != NULL && !IsReplicationOriginSessionSetup(connection))
+	{
+		/*elog(LOG, "After IsReplicationOriginSessionSetup session %s,%d", connection->hostname, connection->port); */
+		StringInfo replicationOriginSessionSetupQuery = makeStringInfo();
+		appendStringInfo(replicationOriginSessionSetupQuery,
+						 "select pg_catalog.replication_origin_session_start_no_publish();");
+		ExecuteCriticalRemoteCommand(connection,
+									 replicationOriginSessionSetupQuery->data);
+		connection->isReplicationOriginSessionSetup = true;
+
+		/*elog(LOG, "After Setting up remote replication origin session %s,%d", connection->hostname, connection->port); */
+	}
+}
+
+
+/*
+ * ResetReplicationOriginRemoteSession resets the replication origin session in a
+ * remote node.
+ */
+void
+ResetReplicationOriginRemoteSession(MultiConnection *connection)
+{
+	if (connection != NULL && connection->isReplicationOriginSessionSetup)
+	{
+		/*elog(LOG, "Resetting remote replication origin session %s,%d", connection->hostname, connection->port); */
+
+		/* Reset the replication origin in remote session*/
+		StringInfo replicationOriginSessionResetQuery = makeStringInfo();
+		appendStringInfo(replicationOriginSessionResetQuery,
+						 "select pg_catalog.replication_origin_session_end_no_publish();");
+		ExecuteCriticalRemoteCommand(connection,
+									 replicationOriginSessionResetQuery->data);
+		connection->isReplicationOriginSessionSetup = false;
+	}
+}
+
+
+/*
+ * SetupReplicationOriginLocalSession sets up a new replication origin session in a
+ * local session.
+ */
+void
+SetupReplicationOriginLocalSession(void)
+{
+	/*elog(LOG, "Setting up local replication origin session"); */
+	if (!IsReplicationOriginSessionSetup(NULL))
+	{
+		repl_origin_local_session_start_no_publish();
```

Comment:
How do we reset the local replication origin ID in case of error? we discussed using a memory context callback, though could also use transaction callbacks.

Reply:
I will add the memory context and transaction callbacks to Reset the Replication Origin.

Reply:
Added callback functions for Memorycontext destroy event, for two places: multi_copy.c and worker_shard_copy.c, but not able to find a proper Memory context for create_distribute_table..c and local_memory_copy.c.  This is for resetting of replication origin for local session only. For remote sessions, the replication origin is reset automatically, when the session is closed.